Mathlib/Data/Fin/Tuple/Sort.lean

```diff
@@ -109,12 +111,52 @@ theorem monotone_sort (f : Fin n → α) : Monotone (f ∘ sort f) := by
 
 end Tuple
 
+theorem Fintype.card_fin_lt_nat (m g : ℕ) (h : g ≤ m) : Fintype.card {i : Fin m // i < g } = g := by
+  conv_rhs => rw [← Fintype.card_fin g]
+  apply Fintype.card_congr
+  exact ⟨ fun x => ⟨x, x.prop⟩, fun x => ⟨⟨x, (lt_of_lt_of_le (Fin.is_lt x) h)⟩, x.prop⟩,
+    fun x => by simp, fun x => by simp⟩
+
 namespace Tuple
 
 open List
 
 variable {n : ℕ} {α : Type*}
 
+/-- A sorted tuple with `m` elements and exactly `Fintype.card {i // f i ≤ a}` less than `a`, has
+the elements at the start, and vice versa -/
+theorem lt_card_le_iff_apply_le_of_monotone [PartialOrder α] [DecidableRel (α := α) LE.le]
+    {m : ℕ} (f : Fin m → α) (a : α) (h_sorted : Monotone f) (j : Fin m) :
+    j < Fintype.card {i // f i ≤ a} ↔ f j ≤ a := by
+  suffices h1 : ∀ k : Fin m, (k < Fintype.card {i // f i ≤ a}) → f k ≤ a
+  refine ⟨h1 j, ?_⟩
+  · intro h
+    by_contra' hc
+    let p (x : Fin m) := f x ≤ a
+    let q (x : Fin m) := x < Fintype.card {i // f i ≤ a}
+    let q' (x : {i // f i ≤ a}) := q x
+    have hw : 0 < Fintype.card {j : {x : Fin m // f x ≤ a} // ¬q' j} :=
+      Fintype.card_pos_iff.2 (Nonempty.intro ⟨⟨j, h⟩, not_lt.2 hc⟩)
+    have he := Fintype.card_congr <| Equiv.sumCompl <| q'
+    have h4 := (Fintype.card_congr (@Equiv.subtypeSubtypeEquivSubtype _ p q (h1 _)))
+    rw [Fintype.card_sum, h4, Fintype.card_fin_lt_nat, add_right_eq_self] at he
+    · exact hw.ne he.symm
+    conv_rhs => rw [← Fintype.card_fin m]
+    exact Fintype.card_subtype_le _
+  · intro _ h
+    contrapose! h
+    rw [← Fin.card_Iio, Fintype.card_subtype]
+    refine Finset.card_mono (fun i => Function.mtr ?_)
+    simp_rw [Finset.mem_filter, Finset.mem_univ, true_and, Finset.mem_Iio]
+    intro hij hia
+    apply h
+    exact (h_sorted (le_of_not_lt hij)).trans hia
```

Comment:
Most of this is just nitpicks but the original proof did commit two minor stylistic sins, namely:
 1. It did not capture the `suffices` goal using `· ` (instead immediately working on the head goal with `refine`)
 2. It spawned a subgoal in the rewrite using `Fintype.card_fin_lt_nat` (which I have renamed to `Fintype.card_fin_lt_of_le`). Spawning subgoals during a rewrite like this makes proofs harder to follow and maintain so it is better to create the require hypothesis in advance (as I have done with `h_le`).
```suggestion
  · refine ⟨h1 j, fun h ↦ ?_⟩
    by_contra' hc
    let p : Fin m → Prop := fun x ↦ f x ≤ a
    let q : Fin m → Prop := fun x ↦ x < Fintype.card {i // f i ≤ a}
    let q' : {i // f i ≤ a} → Prop := fun x ↦ q x
    have hw : 0 < Fintype.card {j : {x : Fin m // f x ≤ a} // ¬ q' j} :=
      Fintype.card_pos_iff.2 ⟨⟨⟨j, h⟩, not_lt.2 hc⟩⟩
    apply hw.ne'
    have he := Fintype.card_congr <| Equiv.sumCompl <| q'
    have h4 := (Fintype.card_congr (@Equiv.subtypeSubtypeEquivSubtype _ p q (h1 _)))
    have h_le : Fintype.card { i // f i ≤ a } ≤ m := by
      conv_rhs => rw [← Fintype.card_fin m]
      exact Fintype.card_subtype_le _
    rwa [Fintype.card_sum, h4, Fintype.card_fin_lt_of_le h_le, add_right_eq_self] at he
  intro _ h
  contrapose! h
  rw [← Fin.card_Iio, Fintype.card_subtype]
  refine Finset.card_mono (fun i => Function.mtr ?_)
  simp_rw [Finset.mem_filter, Finset.mem_univ, true_and, Finset.mem_Iio]
  intro hij hia
  apply h
  exact (h_sorted (le_of_not_lt hij)).trans hia
```
Further optimisations are still possible but I think the above will do.